mesa/space.py

```diff
@@ -81,237 +62,207 @@ def __init__(self, width, height, torus):
+            for dy in range(-radius, radius + 1):
+                x, y = self._torus_adj((pos[0] + dx, pos[1] + dy))
+                # If a cell is empty and empty cell should not be returned, skip
+                if (self.grid[x][y] == self.empty_value) and (not include_empty):
+                    continue
+                # Skip coordinates that are outside manhattan distance
+                if abs(dx) + abs(dy) > radius:
+                    break
+                if (get_agents):
+                    neighbors.add((self.grid[x][y], (x, y)))
+                else:
+                    neighbors.add((x, y))
+
+    def _torus_adj(self, pos):
+        """ Convert coordinate, handling torus looping. """
+        if self.torus:
+            x, y = pos[0] % self.width, pos[1] % self.height
+            return x, y
+        else:
+            if (not 0 <= pos[0] < self.width) or (not 0 <= pos[1] < self.height):
+                raise IndexError(
+                    "Coordinates out of bounds. Grid is non-toroidal.")
 
-    def move_agent(self, agent, pos):
+    def move_agent(self, agent, pos, replace=False):
         """
         Move an agent from its current position to a new position.
 
         Args:
             agent: Agent object to move. Assumed to have its current location
                    stored in a 'pos' tuple.
             pos: Tuple of new position to move the agent to.
-
+            replace: if True, replace the possibly existed agent at `pos` with `agent`
         """
-        pos = self.torus_adj(pos)
+        pos = self._torus_adj(pos)
         self._remove_agent(agent.pos, agent)
-        self._place_agent(pos, agent)
+        self._place_agent(pos, agent, replace)
         agent.pos = pos
 
-    def place_agent(self, agent, pos):
-        """ Position an agent on the grid, and set its pos variable. """
-        self._place_agent(pos, agent)
-        agent.pos = pos
+    def place_agent(self, agent, pos=("random", "random"), replace=False):
+        """ Position an agent on the grid, and set its pos variable. 
+
+        Args:
+            agent: agent to place at `pos`
+            pos: coordinates of the grid (x, y)
+            replace: if True, replace the possibly existed agent at `pos` with `agent`
+        """
+        if (pos[0] == "random") or (pos[1] == "random"):
+            coords = self._pick_random_position(agent, pos)
+        self._place_agent(coords, agent, replace)
+        agent.pos = coords
 
-    def _place_agent(self, pos, agent):
+    def _place_agent(self, pos, agent, replace):
         """ Place the agent at the correct location. """
         x, y = pos
-        self.grid[x][y] = agent
-        if pos in self.empties:
+        old_agent = self.grid[x][y]
+
+        if old_agent == self.empty_value:
+            self.grid[x][y] = set(agent) if self.multigrid else agent
             self.empties.remove(pos)
+        else:
+            if replace:
+                old_agent.pos = None
```

Comment:
I think this will fail for multigrid? As old_agent would be a set. It is also not clear how "replace" should work for a multigrid. Should it replace a single agent? All agents? Or can't it be used with MultiGrid?